### PR TITLE
Fixed amcl related lifecycle issues

### PR DESCRIPTION
**Pull request by mhpanah**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | () |
| Primary OS tested on | (Ubuntu 18.04) |
| Robotic platform tested on | (Gazebo and TB3) |

---

## Description of contribution in a few bullet points

- Originally at startup, amcl publishes (0,0,0) for position since it doesn't know it's initial position.  We updated AMCL to not publish a pose or update transform if it has not been globally localized. AMCL can globally be localized at startup by either setting initial pose manually from user, getting the last pose from a file or server, or by auto-localization. 
  - This change, causes the `costmap_2d_ros` to get stuck in `on_activate` state.  Per team discussion, once we merge the lifecycle branch onto master we will revisit this issue.

- Currently, the `initial pose subscriber` is set to have the default QOS settings. With this setting, the `localization system test` was failing due to publishing the initial pose before amcl node gets active.  To fix this issue, I modified the localization test to periodically send the initial pose. However, after merging lifecycle branch to master we need to change the amcl's `initial pose subscriber` QOS to `transient_local().reliable()`. This won't work now, because in rviz `transient_local` is not set. 

- Enabled obstacle layer

- Changing rcl_info to rcl_debug for publishing velocity

---

## Future work that may be required in bullet points

-Revisit `costmap_2d_ros` activate state and decide if we should block amcl's activation or costmap.
-Change AMCL `initial pose subscriber` QOS to `transient_local().reliable()`
-Factor out `checkElapsedTime()` to `nav2_util` so that it can be used elsewhere

---

### CODE REVIEW

File: nav2_bringup/launch/nav2_params.yaml [hunk 1]
```diff
@@ -95,7 +95,7 @@ local_costmap:
       robot_radius: 0.17
       inflation_layer.cost_scaling_factor: 3.0
       obstacle_layer:
-        enabled: False
+        enabled: True
```
Comment (orduno) [inline nav2_bringup/launch/nav2_params.yaml]: Minor observation: obstacle layer is already enabled by #762. In addition to changing this parameter, QoS setting for laser scan inside `ObstacleLayer` has to be set to `sensor_data`.

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp [hunk 2]
```diff
@@ -137,6 +137,8 @@ class AmclNode : public nav2_lifecycle::LifecycleNode
     double & x, double & y, double & yaw,
     const rclcpp::Time & sensor_timestamp, const std::string & frame_id);
   std::atomic<bool> first_pose_sent_;
+  std::atomic<bool> amcl_node_ready_;
+  std::atomic<bool> first_laser_scan_rec_;
```
Comment (bpwilcox) [inline nav2_amcl/include/nav2_amcl/amcl_node.hpp]: Where/how is this used? I don't see it being set or used in the code anywhere

File: nav2_bringup/launch/nav2_params.yaml (hunk 1 above)
Comment (mhpanah) [reply nav2_bringup/launch/nav2_params.yaml]: Will rebase.

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp (hunk 2 above)
Comment (mhpanah) [reply nav2_amcl/include/nav2_amcl/amcl_node.hpp]: `first_laser_scan_rec_` should be removed.

**Comment by mjeronimo** [comment]
Please consider that AMCL should be able to go through the DEACTIVATE, CLEANUP transitions and then be reactivated with CONFIGURE and ACTIVATE and be up and running again. The new state variables introduced should be properly reset, if necessary. However, I think that AMCL should remember that it has issued the initial pose even after it has gone through this cycle.

File: nav2_amcl/src/amcl_node.cpp [hunk 3]
```diff
@@ -363,25 +375,38 @@ AmclNode::nomotionUpdateCallback(
...
+      "but AMCL is not yet in the active state");
+    return;
+  }
+  handleInitialPose(msg);
+}
+
+void
+AmclNode::handleInitialPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  // initOdometry();
```
[the remark this answers is not in the document]
Comment (mhpanah) [reply nav2_amcl/src/amcl_node.cpp]: Fixed.

**Comment by mkhansen-intel** [comment]
Please don't merge until it is passing CI

**Comment by mkhansen-intel** [comment]
Since test is disabled, this should be OK

**Pull request merged**